docs/source/design.rst

```diff
@@ -1,7 +1,201 @@
-.. _internals:
+.. _design:
 
-*********************
-Ibis design internals
-*********************
+Design
+======
 
-More to come here.
+
+.. _primary_goals:
+
+Primary Goals
+-------------
+
+#. Type safety
+#. Expressiveness
+#. Composability
+#. Familiarty
+
+.. _flow_of_execution:
+
+Flow of Execution
+-----------------
+
+#. User writes expression
+#. Each method or function call builds a new expression
+#. Expressions are type checked as you create them
+#. Some expressions have some optimizations that happen as the user builds them
+   (EXAMPLE)
+#. Some optimizations happen at compile time? (EXAMPLE)
+#. Expressions are compiled
+#. The SQL string that generated by the compiler is sent to the database and
+   executed (this step is skipped for the Pandas backend)
+#. The database returns some data that is then turned into a pandas DataFrame
+   by ibis
+
+.. _expressions:
+
+Expressions
+-----------
+
+The main user-facing component of ibis is expressions. The base class of all
+expressions in ibis is the :class:`~ibis.expr.types.Expr` class.
+
+Expressions provide the user facing API, defined in ``ibis/expr/api.py``
+
+.. _type_system:
+
+Type System
+~~~~~~~~~~~
+
+Ibis's type system consists of a set of rules for specifying the types of
+inputs to :class:`~ibis.expr.types.Node` subclasses. Upon construction of a
+:class:`~ibis.expr.types.Node` subclass, ibis performs validation of every
+input to the node based on the rule that was used to declare the input.
+
+Rules are defined in ``ibis/expr/rules.py``
+
+.. _expr_class:
+
+The :class:`~ibis.expr.types.Expr` class
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Expressions are a thin but important abstraction over operations, containing
+only type information and shape information, i.e., whether they are tables,
+columns, or scalars.
+
+Examples of expressions include :class:`~ibis.expr.types.Int64Column`,
+:class:`~ibis.expr.types.StringScalar`, and
+:class:`~ibis.expr.types.TableExpr`.
+
+Here's an example of each type of expression:
+
+.. code-block:: python
```

Comment:
Could do IPython directive (with graphviz disabled)

Reply:
This won't show them anyways, since I'm not repring the expression, just the type.